File: src/leetcode/python/FindMedianFromDataStream_295.py
from heapq import *


class FindMedianFromDataStream_295(object):
    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.left = []
        self.right = []

    def addNum(self, num):
        """
        Adds a num into the data structure.
        :type num: int
        :rtype: void
        """
        if len(self.left) == len(self.right):
            heappush(self.right, -heappushpop(self.left, -num))
        else:
            heappush(self.left, -heappushpop(self.right, num))

    def findMedian(self):
        """
        Returns the median of current data stream
        :rtype: float
        """
        if len(self.left) == len(self.right):
            return float(self.right[0] - self.left[0]) / 2
        return float(self.right[0])

File: src/leetcode/java/FindMedianFromDataStream_295.java
package leetcode.java;

import java.util.Collections;
import java.util.PriorityQueue;

public class FindMedianFromDataStream_295 {
  private PriorityQueue<Integer> leftHeap = new PriorityQueue<>(Collections.reverseOrder());
  private PriorityQueue<Integer> rightHeap = new PriorityQueue<>();
  boolean even = true;

  // Adds a number into the data structure.
  public void addNum(int num) {
    if (even) {
      rightHeap.offer(num);
      leftHeap.offer(rightHeap.poll());
    } else {
      leftHeap.offer(num);
      rightHeap.offer(leftHeap.poll());
    }
    even = !even;
  }

  // Returns the median of current data stream
  public double findMedian() {
    if (even) {
      return (double) (leftHeap.peek() + rightHeap.peek()) / 2;
    }
    return leftHeap.peek();
  }
}
